spring-integration-zip/src/main/java/org/springframework/integration/zip/splitter/UnZipResultSplitter.java

```diff
@@ -21,32 +21,51 @@
 import java.util.Map;
 
 import org.apache.commons.io.FilenameUtils;
+
 import org.springframework.integration.file.FileHeaders;
 import org.springframework.integration.support.MessageBuilder;
 import org.springframework.integration.zip.ZipHeaders;
 import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
 
 /**
  *
  * @author Gunnar Hillert
+ * @author Andriy Kryvtsun
  * @since 1.0
  *
  */
 public class UnZipResultSplitter {
 
-	public List<Message<Object>> splitUnzippedMap(Map<String, Object> unzippedEntries) {
-
-		final List<Message<Object>> messages = new ArrayList<Message<Object>>(unzippedEntries.size());
+	public List<Message<Object>> splitUnzippedMap(Message<Map<String, Object>> message) {
+		return createMessages(message.getPayload(), message.getHeaders());
+	}
 
+	private List<Message<Object>> createMessages(Map<String, Object> unzippedEntries, MessageHeaders headers) {
+		List<Message<Object>> messages = createList(unzippedEntries.size());
 		for (Map.Entry<String, Object> entry : unzippedEntries.entrySet()) {
-			final String path = FilenameUtils.getPath(entry.getKey());
-			final String filename = FilenameUtils.getName(entry.getKey());
-			final Message<Object> splitMessage = MessageBuilder.withPayload(entry.getValue())
-					.setHeader(FileHeaders.FILENAME, filename)
-					.setHeader(ZipHeaders.ZIP_ENTRY_PATH, path).build();
-			messages.add(splitMessage);
+			messages.add(createMessage(entry, headers));
 		}
 		return messages;
 	}
 
+	private static List<Message<Object>> createList(int initialCapacity) {
+		return new ArrayList<Message<Object>>(initialCapacity);
+	}
+
+	private static Message<Object> createMessage(Map.Entry<String, Object> entry, Map<String, ?> headers) {
+		return MessageBuilder.withPayload(entry.getValue())
+				.setHeader(FileHeaders.FILENAME, getName(entry))
+				.setHeader(ZipHeaders.ZIP_ENTRY_PATH, getPath(entry))
+				.copyHeaders(headers)
+				.build();
```

Comment:
@artembilan pls, clarify proposed change affect another methods too. So are you want to change finally signature of the `public` method `List<Message<Object>> splitUnzippedMap(...)` to `List<MessageBuilder<Object>> splitUnzippedMap(...)`?